File: 01.02.2023/py/main.py
# Write a function which formats a duration, given as a number of seconds, in a human-friendly way.
# The function must accept a non-negative integer. If it is zero, it just returns "now".
# Otherwise, the duration is expressed as a combination of years, days, hours, minutes and seconds.


class WrongType(Exception):
    pass

# Solution one
def format_duration(seconds):
    if seconds == 0:
        return 'now'

    units = [
        (60 * 60 * 24 * 365, 'year'),
        (60 * 60 * 24, 'day'),
        (60 * 60, 'hour'),
        (60, 'minute'),
        (1, 'second'),
    ]

    components = []

    for unit_time, unit_name in units:
        if seconds >= unit_time:
            count = seconds // unit_time
            seconds = seconds % unit_time

            if count > 1:
                unit_name = unit_name + 's'

            components.append(f'{count} {unit_name}')

    if len(components) > 1:
        return ' and '.join([', '.join(components[:-1]), components[-1]])
    else:
        return components[0]


import cProfile
import time
from timeit import default_timer as timer


if __name__ == '__main__':
    cProfile.run('format_duration(1)')
    cProfile.run('format_duration(62)')

    start_time = time.time()
    format_duration(3662)
    end_time = time.time()
    print('', end_time - start_time)

File: 01.02.2023/js/main.js
// Write a function which formats a duration, given as a number of seconds, in a human-friendly way.
// The function must accept a non-negative integer. If it is zero, it just returns "now".
// Otherwise, the duration is expressed as a combination of years, days, hours, minutes and seconds.


/* Solution one */
export const formatDuration = (seconds) => {
    if (seconds === 0) return 'now';

    const units = [
        {value: 31536000, name: 'year'},
        {value: 86400, name: 'day'},
        {value: 3600, name: 'hour'},
        {value: 60, name: 'minute'},
        {value: 1, name: 'second'},
    ];

    let result = '';

    for (let i = 0; i < units.length; i++) {
        if (seconds >= units[i].value) {
            let count = Math.floor(seconds / units[i].value);

            seconds = seconds % units[i].value;
            result += `${count} ${units[i].name}` + (count > 1 ? 's' : '') + ', ';
        }
    }

    return result.slice(0, -2).replace(/,([^,]*)$/, ' and$1');
}
